### ISSUE TITLE
0.14.2 Failure: Cannot find "ocamlbuild.cmo" in ocamlbuild -where directory.

### ISSUE THREAD

**Issue by SnarkBoojum** [ocaml/ocamlbuild#321]
Ocamlbuild 0.14.2 compiles successfully with "make", but at the "make test" stage, several of the tests fail with a similar error:
```
Warning: Won't be able to compile a native plugin
Failure: Cannot find "ocamlbuild.cmo" in ocamlbuild -where directory.
```

Version 0.14.1 is installed on my system, and I'm using OCaml 4.13.1.

I tested:
1. "make && make test" in 0.14.1's source directory works on my system ;
2. "make && make test" in 0.14.2's source directory fails on my system ;
3. "make && make test" in 0.14.1's source directory works in a chroot without an installed ocamlbuild ;
4. "make && make test" in 0.14.2's source directory fails in a chroot without an installed ocamlbuild.

So the problem is really something new in 0.14.2.

**Comment by gasche** [comment]
Here is my understanding of the issue:

- the bug was introduced in #319, which implemented support for OCaml 5.0
- the issue is that, after #319, `ocamlbuild.cmo` is compiled in a different directory from other object files (`bin/` instead of `src/`), while our test infrastructure assumes that all library/plugin files can be found in the same directory that is passed to ocamlbuild using the `-install-lib-dir` option
- `-install-lib-dir` sets a single constant, I cannot pass several directories there
- the reason why #319 moved is to solve a small issue with the Makefile-based build system, that would build all source files with `-for-pack Ocamlbuild_pack`, even though some files were actually not packed. This used to be morally incorrect but work fine, and it is now explicitly rejected by OCaml 5.x ( see https://github.com/dbuenzli/mtime/issues/45 ).

There are two fixes that would make sense:
1. we could consider adding support for providing several "library directories" to ocamlbuild at runtime; those are used when linking plugins
2. or we could change the build system to store all build objects in a single directory, finding another workaround to the -for-pack issue

Approach (1) is nicer on paper but it requires more effort, and is more invasive/risky. Approach (2) is somewhat ugly but should make no observable change to ocamlbuild users (after install).

Given the fact that ocamlbuild is now in maintenance-only mode, I would vote for approach (2).

**Comment by SnarkBoojum** [comment]
Or perhaps I should wait until I upgrade to OCaml 5.1 before I upgrade to ocamlbuild 0.14.2 ?

**Comment by gasche** [comment]
Yes, 0.14.1 should be fine if your OCaml version is 5.0 or older.

**Comment by gasche** [comment]
(But I still need to fix this bug in the current codebase. Thanks for the report!)

**Comment by jonahbeckford** [comment]
I just ran into this on Windows and filed a bug at https://github.com/diskuv/dkml-installer-ocaml/issues/70

I have two `diskuvocaml*.patch` patches in https://github.com/diskuv/diskuv-opam-repository/tree/main/packages/ocamlbuild/ocamlbuild.0.14.0/files that fix the problem, and forgot to re-apply them for ocamlbuild.0.14.2.

The root problem (fixed by `diskuvocaml.patch`) is that backslashes are interpreted as unescaped:

```make
# file: Makefile.config
PREFIX=
BINDIR=Z:\source\dkml\build\_deps\dkml-component-opam-src\_opam\bin
LIBDIR=Z:\source\dkml\build\_deps\dkml-component-opam-src\_opam\lib
MANDIR=Z:\source\dkml\build\_deps\dkml-component-opam-src\_opam\man
```

```ocaml
# file: src/ocamlbuild_config.ml
let bindir = "Z:\source\dkml\build\_deps\dkml-component-opam-src\_opam\bin"
let libdir = "Z:\source\dkml\build\_deps\dkml-component-opam-src\_opam\lib"
let ocaml_libdir = "C:/Users/beckf/AppData/Local/Programs/DKMLNA~1/lib/ocaml"
let libdir_abs = "Z:/source/dkml/build/_deps/dkml-component-opam-src/_opam/lib"
```

Additionally, the Makefiles do not support spaces, which is fixed by `diskuvocaml2.patch`. I'll send out PRs and refer to this issue.

**Comment by gasche** [comment]
@jonahbeckford I don't understand the relation between the patches you are showing and the issue here, that I think is related to the ocamlbuild build system and testsuite layout that breaks `make test`. Could it be that you see the same error message, but in a different situation (not only when running `make test`, but also after ocamlbuild is installed)? If so, I think we should open a separate issue or stick to the PRs you are opening (thanks!), to not muddy the conversation here.

**Comment by hhugo** [comment]
This can be closed (by #328)
